Custom events for core client communication (#746)
SAP/luigi: docs/communication.md

# Communication

## Custom messages

> Luigi Core and Luigi Client can exchange custom messages in two possible ways.

Reviewer: ```suggestion
Luigi Core and Luigi Client can exchange custom messages in both directions.
```

> #### Luigi Client sends a custom message to Luigi Core

Reviewer: ```suggestion
### Luigi Client to Luigi Core
```

> For Luigi Client, you need to use the [*sendCustomMessage*](luigi-client-api.md#sendCustomMessage) method from Client API.

Reviewer: ```suggestion
For Luigi Client to send messages to Luigi Core, use the [*sendCustomMessage*](luigi-client-api.md#sendCustomMessage) method from Client API.
```

> For Luigi Core, you need to set something similar to the following on the root level of your Luigi configuration object:

Reviewer: ```suggestion
For Luigi Core to receive custom messages, define a configuration similar to the following at the root level of your Luigi configuration object:
```

Reviewer 2: I will apply your suggestion, but I will use "process" instead of "receive", since the message actually always arrives to Core.

Reviewer: Sounds good to me :)

```javascript
{
  ...
  communication: {
    customMessagesListeners: {
      'my-custom-message.update-top-nav': () => {
        Luigi.navigation().updateTopNavigation();
      }
    }
  }
  ...
}
```

> where the key `'my-custom-message.update-top-nav` is the message id and the value is the listener function for the custom message. The listener receives the following input parameters:

Reviewer: ```suggestion
where the `'my-custom-message.update-top-nav` key is the message id, and the value is the listener function for the custom message. The listener receives the following input parameters:
```

> - **customMessage** the [*message*](luigi-client-api.md#sendCustomMessage) sent by the client.

Reviewer: ```suggestion
- **customMessage** the [*message*](luigi-client-api.md#sendCustomMessage) sent by Luigi Client.
```

> - **microfrontend** a microfrontend object as specified [here](luigi-core-api.md#getMicrofrontends).

Reviewer: ```suggestion
- **microfrontend** a micro frontend object as specified [here](luigi-core-api.md#getMicrofrontends).
```

- **navigation node** a [navigation node object](navigation-parameters-reference.md#Node-parameters).

> #### Luigi Core sends a custom message to Luigi Client

Reviewer: ```suggestion
### Luigi Core to Luigi Client
```

> For Luigi Core, you need to use the [*customMessages*](luigi-core-api.md#customMessages) section from Core API. It is possible to send a custom message to all the rendered micro frontends, or to a specific one. For the latter, you can make use of the Core API [*elements*](luigi-core-api.md#elements) methods to retrieve the micro frontends and select the one you want to send the custom message to.

Reviewer: ```suggestion
For Luigi Core to send messages, use the [*customMessages*](luigi-core-api.md#customMessages) section from Core API. You can send a custom message to all rendered micro frontends, or to a specific one. For the latter, use the Core API [*elements*](luigi-core-api.md#elements) methods to retrieve micro frontends and select the one you want to send the custom message to.
```

> For Luigi Client, you can add and remove message listeners as described [here](luigi-client-api.md#addCustomMessageListener).

Reviewer: ```suggestion
For Luigi Client to receive the message, add and remove message listeners as described [here](luigi-client-api.md#addCustomMessageListener).
```
